### PR TITLE
sDDF block

### PR DESCRIPTION
**Pull request by erichchan999**

### CODE REVIEW

File: include/sddf/block/blk_shared_queue.h [hunk 1]
```diff
@@ -0,0 +1,342 @@
...
+typedef enum blk_response_status {
+    SUCCESS,
+    SEEK_ERROR
+} blk_response_status_t;
+
+/* Request struct contained in request queue */
+typedef struct blk_request {
+    blk_request_code_t code; /* request code */
+    uintptr_t addr; /*  */
+    uint32_t sector; /* sector number to read/write */
```
Comment (wom-bat) [inline include/sddf/block/blk_shared_queue.h]: # blocks not sectors

File: include/sddf/block/blk_shared_queue.h [hunk 2]
```diff
@@ -0,0 +1,342 @@
...
+    uint16_t count; /* number of blocks allocated for corresponding request */
+    uint16_t success_count; /* number of blocks successfully read/written */
+    uint32_t id; /* stores corresponding request ID */
+} blk_response_t;
+
+/* Circular buffer containing requests */
+typedef struct blk_req_queue {
+    uint32_t write_idx;
+    uint32_t read_idx;
+    uint32_t size; /* number of buffers in request queue */
```
Comment (wom-bat) [inline include/sddf/block/blk_shared_queue.h]: number of slots, not buffers

File: include/sddf/block/blk_shared_queue.h [hunk 3]
```diff
@@ -0,0 +1,342 @@
...
+ * @param response pointer to response queue in shared memory.
+ * @param buffer_init 1 indicates the read and write indices in shared memory need to be initialised.
+ *                    0 indicates they do not. Only one side of the shared memory regions needs to do this.
+ * @param request_size number of buffers in request queue.
+ * @param response_size number of buffers in response queue.
+ */
+void blk_queue_init(blk_queue_handle_t *queue_handle,
+                        blk_req_queue_t *request,
+                        blk_resp_queue_t *response,
+                        int buffer_init,
```
Comment (wom-bat) [inline include/sddf/block/blk_shared_queue.h:98]: bool ?

**Review by wom-bat — DISMISSED** [review]
This now looks good as a starting point.
I expect as we use it it'll evolve.

File: include/sddf/block/blk_shared_queue.h [hunk 4]
```diff
@@ -0,0 +1,342 @@
...
+
+#pragma once
+
+#include <stdint.h>
+#include <stddef.h>
+#include <stdbool.h>
+#include "fence.h"
+
+/* Number of buffers the request queue is configured to have. */
+#define BLK_REQ_QUEUE_SIZE 1024
```
Comment (abrandnewusername) [inline include/sddf/block/blk_shared_queue.h:15]: where does the number come from?

File: include/sddf/block/blk_shared_queue.h [hunk 5]
```diff
@@ -0,0 +1,342 @@
...
+/* Number of buffers the response queue is configured to have. */
+#define BLK_RESP_QUEUE_SIZE 1024
+
+typedef struct blk_storage_info {
+    char serial_number[64];
```
Comment (abrandnewusername) [inline include/sddf/block/blk_shared_queue.h:20]: where does `64` come from?

File: include/sddf/block/blk_shared_queue.h [hunk 6]
```diff
@@ -0,0 +1,342 @@
...
+    bool read_only;
+    bool ready;
+    uint16_t blocksize;
+    uint16_t queue_depth;
+    uint16_t cylinders, heads, blocks; /* geometry to guide FS layout */
+    uint64_t size; /* number of blocksize units */
+} blk_storage_info_t;
+
+/* Request code for block */
+typedef enum blk_request_code {
```
Comment (abrandnewusername) [inline include/sddf/block/blk_shared_queue.h:30]: why do we need to define request code in the library in stead of letting the users to decide what they need to have?

File: include/sddf/block/blk_shared_queue.h [hunk 7]
```diff
@@ -0,0 +1,342 @@
...
+                                        uintptr_t addr,
+                                        uint32_t block_number,
+                                        uint16_t count,
+                                        uint32_t id)
+{
+    if (blk_req_queue_full(queue_handle)) {
+        return -1;
+    }
+
+    queue_handle->req_queue->buffers[queue_handle->req_queue->write_idx % queue_handle->req_queue->size].code = code;
```
Comment (abrandnewusername) [inline include/sddf/block/blk_shared_queue.h:197]: Might be worth it to have a macro for `queue_handle->req_queue->buffers[queue_handle->req_queue->write_idx % queue_handle->req_queue->size]`

File: include/sddf/block/blk_shared_queue.h [hunk 8]
```diff
@@ -0,0 +1,342 @@
...
+    return 0;
+}
+
+/**
+ * Set the plug of the request queue to true.
+ *
+ * @param queue_handle queue handle containing request queue to check for plug.
+*/
+static inline void blk_req_queue_plug(blk_queue_handle_t *queue_handle) {
+    queue_handle->req_queue->plugged = true;
```
Comment (abrandnewusername) [inline include/sddf/block/blk_shared_queue.h:320]: `plugged` seems to be a boolean that only gets set and checked by the users, and doesn't have anything to do with the implementation of the shared queue library despite the fact that the comment describes that it "prevents requests from being dequeued when plugged".

Comment (erichchan999) [reply include/sddf/block/blk_shared_queue.h:320]: The plugged is for optimisation where the driver is not meant to handle any subsequent requests when plugged. This functionality has to be communicated through some shared region + signalling protocol regardless, so it also make sense to put it in the shared library.

File: include/sddf/block/blk_shared_queue.h (hunk 5 above)
Comment (erichchan999) [reply include/sddf/block/blk_shared_queue.h:20]: most serial numbers should fit within 64 characters, will leave a comment

File: include/sddf/block/blk_shared_queue.h (hunk 4 above)
Comment (erichchan999) [reply include/sddf/block/blk_shared_queue.h:15]: typo, "buffers" should change to "slots" here, indicates the number of slots in the request queue. Should be configurable to maximise performance

File: include/sddf/block/blk_shared_queue.h (hunk 6 above)
Comment (erichchan999) [reply include/sddf/block/blk_shared_queue.h:30]: the client + server will need to agree on some interface regardless. The idea of specifying the type of requests in the shared library is that they should cover all use cases for block devices

File: blk/libblksharedqueue/Makefile [hunk 9]
```diff
@@ -22,11 +22,11 @@ CFLAGS := \
 	-I$(SDDF)/include \
 	-I$(SDDF)/util/include
 
-all: $(BUILD_DIR)/sddf_block_shared_queue.o
```
Comment (Ivan-Velickovic) [inline blk/libblksharedqueue/Makefile]: The object file should still retain the prefix to not clash with anything.

Comment (Ivan-Velickovic) [reply blk/libblksharedqueue/Makefile]: It's unlikely, but this naming will be consistent with the other ring buffers etc

File: include/sddf/block/blk_shared_queue.h (hunk 6 above)
Comment (wom-bat) [reply include/sddf/block/blk_shared_queue.h]: This is part of the definition of the interface, it has to be here.

File: blk/libblksharedqueue/Makefile [hunk 10]
```diff
@@ -0,0 +1,32 @@
...
+
+ifeq ($(strip $(MICROKIT_INCLUDE)),)
+$(error MICROKIT_INCLUDE must be specified)
+endif
+
+BUILD_DIR ?= .
+
+CC := clang
+TARGET := aarch64-none-elf
+SDDF := ../../
```
Comment (wom-bat) [inline blk/libblksharedqueue/Makefile:15]: Likewise, ?= for SDDF

File: blk/libblksharedqueue/shared_queue.c [hunk 11]
```diff
@@ -0,0 +1,165 @@
+/*
+ * Copyright 2023, UNSW
```
Comment (wom-bat) [inline blk/libblksharedqueue/shared_queue.c:2]: Update to 2024

File: blk/libblksharedqueue/shared_queue.c [hunk 12]
```diff
@@ -0,0 +1,165 @@
...
+        h->req_queue->read_idx = 0;
+        h->req_queue->size = request_size;
+        h->req_queue->plugged = false;
+        h->resp_queue->write_idx = 0;
+        h->resp_queue->read_idx = 0;
+        h->resp_queue->size = response_size;
+    }
+}
+
+bool blk_req_queue_empty(blk_queue_handle_t *h)
```
Comment (wom-bat) [inline blk/libblksharedqueue/shared_queue.c:30]: Move to header, make `static inline`

File: blk/libblksharedqueue/shared_queue.c [hunk 13]
```diff
@@ -0,0 +1,165 @@
...
+{
+    return !((h->req_queue->write_idx - h->req_queue->read_idx) % h->req_queue->size);
+}
+
+bool blk_resp_queue_empty(blk_queue_handle_t *h)
```
Comment (wom-bat) [inline blk/libblksharedqueue/shared_queue.c:35]: Move to header, make `static inline`

File: blk/libblksharedqueue/shared_queue.c [hunk 14]
```diff
@@ -0,0 +1,165 @@
...
+{
+    return !((h->resp_queue->write_idx - h->resp_queue->read_idx) % h->resp_queue->size);
+}
+
+bool blk_req_queue_full(blk_queue_handle_t *h)
```
Comment (wom-bat) [inline blk/libblksharedqueue/shared_queue.c:40]: Move to header, make `static inline`

File: blk/libblksharedqueue/shared_queue.c [hunk 15]
```diff
@@ -0,0 +1,165 @@
...
+{
+    return !((h->req_queue->write_idx - h->req_queue->read_idx + 1) % h->req_queue->size);
+}
+
+bool blk_resp_queue_full(blk_queue_handle_t *h)
```
Comment (wom-bat) [inline blk/libblksharedqueue/shared_queue.c:45]: Move to header, make `static inline`

File: blk/libblksharedqueue/shared_queue.c [hunk 16]
```diff
@@ -0,0 +1,165 @@
...
+{
+    return !((h->resp_queue->write_idx - h->resp_queue->read_idx + 1) % h->resp_queue->size);
+}
+
+int blk_req_queue_size(blk_queue_handle_t *h)
+{
+    return (h->req_queue->write_idx - h->req_queue->read_idx);
+}
+
+int blk_resp_queue_size(blk_queue_handle_t *h)
```
Comment (wom-bat) [inline blk/libblksharedqueue/shared_queue.c:55]: Move to header, make `static inline`

File: blk/libblksharedqueue/shared_queue.c [hunk 17]
```diff
@@ -0,0 +1,165 @@
...
+    *success_count = h->resp_queue->buffers[h->resp_queue->read_idx % h->resp_queue->size].success_count;
+    *id = h->resp_queue->buffers[h->resp_queue->read_idx % h->resp_queue->size].id;
+
+    THREAD_MEMORY_RELEASE();
+    h->resp_queue->read_idx++;
+
+    return 0;
+}
+
+void blk_req_queue_plug(blk_queue_handle_t *h)
```
Comment (wom-bat) [inline blk/libblksharedqueue/shared_queue.c:152]: Make `static inline` in header

File: blk/libblksharedqueue/shared_queue.c [hunk 18]
```diff
@@ -0,0 +1,165 @@
...
+{
+    h->req_queue->plugged = true;
+}
+
+void blk_req_queue_unplug(blk_queue_handle_t *h)
```
Comment (wom-bat) [inline blk/libblksharedqueue/shared_queue.c:157]: Likewise

File: blk/libblksharedqueue/shared_queue.c [hunk 19]
```diff
@@ -0,0 +1,165 @@
...
+{
+    h->req_queue->plugged = false;
+}
+
+bool blk_req_queue_plugged(blk_queue_handle_t *h)
```
Comment (wom-bat) [inline blk/libblksharedqueue/shared_queue.c:162]: And again

**Review by wom-bat — CHANGES_REQUESTED** [review]

File: blk/libblksharedqueue/Makefile [hunk 20]
```diff
@@ -0,0 +1,32 @@
...
+# SPDX-License-Identifier: BSD-2-Clause
+#
+
+ifeq ($(strip $(MICROKIT_INCLUDE)),)
+$(error MICROKIT_INCLUDE must be specified)
+endif
+
+BUILD_DIR ?= .
+
+CC := clang
```
[the remark this answers is not in the document]
Comment (Ivan-Velickovic) [reply blk/libblksharedqueue/Makefile:13]: We can sort this out once we have CI where we test the examples (echo server etc) for both GCC and Clang.

**Review by wom-bat — APPROVED** [review]
Looks good to me.

**Pull request merged**